websocket/src/websocket.cpp

```diff
@@ -572,10 +571,7 @@ static dmExtension::Result OnUpdate(dmExtension::Params* params)
                 conn->m_BufferSize = 0;
             }
 
-            if (conn->m_WasConnected)
-            {
-                HandleCallback(conn, EVENT_DISCONNECTED, 0, conn->m_BufferSize);
-            }
+            HandleCallback(conn, EVENT_DISCONNECTED, 0, conn->m_BufferSize);
```

Comment:
Perhaps not the biggest issue, but the problem I have here is that now you can get a DISCONNECTED event even though you weren't connected? That feels a bit strange to me to be honest.

As this PR is about the logic around the states, it seems we something else missing.

Reply:
> The main problem is there are no way to detect in which state connection was when error occurs, so in there is no clue if callback with EVENT_DISCONNECTED will be invoked after error or not. This PR tries to fix it

Your comment implies that you _did_ get a CONNECTED event. Can you not store this in the app? That way you'll know if you'll get a corresponding disconnected call?

Reply:
Note that I'm not entirely against this PR, it's just that I think we shifted the ambiguity towards a different place instead.

Reply:
Let me explain my thoughts when I tried to fix this.
First I thought that I just simply move setting m_WasConnected to true earlier, right after connection was opened(dmConnectionPool::Dial).
After some time, I acknowledge that this approach just moved ambiguity from one place to another. What state of connection can be treated as "connected"? When dns resolved? socket connected? ssl handshake succeed? websocket handshake succeed? So many questions, so little answers.
So I step back and looked at problem from other side. Lets fire callback when internal state of connection changed, just like EVENT_CONNECTED. That leads to pretty straight forward behavior: if connection internally changes state, it fires callback with that state. Thats all=)  

Yes we internally store connection status, and we can run "disconnect" logic if error happens, but there is no errors specification about what kind of errors could happen and what kind of error can lead to connection failure. That brings more ambiguity on top of existing.

Reply:
> if connection internally changes state, it fires callback with that state.

That sounds reasonable.

I think we're actually missing some state(s), to reflect the behavior. E.g. "uninitalized", so that the errors at the beginning of the connect sequence won't have to go to "disconnected".

Anyways, we can probably deal with that later if it becomes an issue.